samcli/lib/intrinsic_resolver/intrinsic_property_resolver.py

```diff
@@ -0,0 +1,811 @@
+"""
+Process and simplifies CloudFormation intrinsic properties such as FN::* and Ref
+"""
+import copy
+import logging
+
+import base64
+import re
+
+from six import string_types
+
+from samcli.lib.intrinsic_resolver.invalid_intrinsic_exception import InvalidIntrinsicException, \
+    verify_intrinsic_type_list, verify_non_null, verify_intrinsic_type_int, verify_in_bounds, \
+    verify_number_arguments, verify_intrinsic_type_str, verify_intrinsic_type_dict, verify_intrinsic_type_bool, \
+    verify_all_list_intrinsic_type, InvalidSymbolException
+
+LOG = logging.getLogger(__name__)
+
+
+class IntrinsicResolver(object):
+    AWS_INCLUDE = "AWS::Include"
+    SUPPORTED_MACRO_TRANSFORMATIONS = [AWS_INCLUDE]
+    _PSEUDO_REGEX = r'AWS::.*?'
+    _ATTRIBUTE_REGEX = r"[a-zA-Z0-9]*?\.?[a-zA-Z0-9]*?"
+    _REGEX_SUB_FUNCTION = r'\$\{(' + _PSEUDO_REGEX + '||' + _ATTRIBUTE_REGEX + r')\}'
+
+    FN_JOIN = "Fn::Join"
+    FN_SPLIT = "Fn::Split"
+    FN_SUB = "Fn::Sub"
+    FN_SELECT = "Fn::Select"
+    FN_BASE64 = "Fn::Base64"
+    FN_FIND_IN_MAP = "Fn::FindInMap"
+    FN_TRANSFORM = "Fn::Transform"
+    FN_GET_AZS = "Fn::GetAZs"
+    REF = "Ref"
+    FN_GET_ATT = "Fn::GetAtt"
+    FN_IMPORT_VALUE = "Fn::ImportValue"
+
+    SUPPORTED_INTRINSIC_FUNCTIONS = [
+        FN_JOIN,
+        FN_SPLIT,
+        FN_SUB,
+        FN_SELECT,
+        FN_BASE64,
+        FN_FIND_IN_MAP,
+        FN_TRANSFORM,
+        FN_GET_AZS,
+        REF,
+        FN_GET_ATT,
+        FN_IMPORT_VALUE
+    ]
+
+    FN_AND = "Fn::And"
+    FN_OR = "Fn::Or"
+    FN_IF = "Fn::If"
+    FN_EQUALS = "Fn::Equals"
+    FN_NOT = "Fn::Not"
+
+    CONDITIONAL_FUNCTIONS = [
+        FN_AND,
+        FN_OR,
+        FN_IF,
+        FN_EQUALS,
+        FN_NOT
+    ]
+
+    def __init__(self, template=None, symbol_resolver=None):
+        """
+        Initializes the Intrinsic Property class with the default intrinsic_key_function_map and
+        conditional_key_function_map.
+
+        In the future, for items like Fn::ImportValue multiple templates can be provided
+        into the function.
+        """
+        self.template = template
+        self._symbol_resolver = symbol_resolver
+
+        self.intrinsic_key_function_map = self.default_intrinsic_function_map()
+        self.conditional_key_function_map = self.default_conditional_key_map()
+
+    @property
+    def template(self):
+        return self._template
+
+    @template.setter
+    def template(self, template):
+        """
+        Set the attributes from the given template. The template is copied to avoid risking mutation of the original
+        template.
+
+        This is defined outside init for readability purposes.
+        :param template: the template dict to parse
+        """
+        # pylint: disable-msg=attribute-defined-outside-init
```

Comment:
Should these be defaulted in `__init__` and then overwritten?

Reply:
All the template initialization has been moved up to the __init__